File: Algorithms/0429_N-ary_Tree_Level_Order_Traversal/Python/N-ary_Tree_Level_Order_Traversal_Solution_1.py

# Space: O(n)
# Time: O(n)

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
import collections


class Solution:
    def levelOrder(self, root):
        if root is None: return []
        res = collections.defaultdict(list)
        res_list = []

        queue = [[0, root]]
        while queue:
            cur_level, cur_root = queue.pop(0)
            res[cur_level].append(cur_root.val)
            if cur_root.children:
                for each_child in cur_root.children:
                    queue.append([cur_level + 1, each_child])

        for i in range(len(res)):
            res_list.append(res[i])
        return [res[i] for i in range(len(res))]

File: Algorithms/0429_N-ary_Tree_Level_Order_Traversal/Java/N-aryTreeLevelOrderTraversalSolution1.java

// Space: O(n)
// Time: O(n)

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    private Map<Integer, List<Integer>> cache;
    private List<List<Integer>> res;

    public List<List<Integer>> levelOrder(Node root) {
        if (root == null) return new ArrayList<>();
        cache = new HashMap<>();
        res = new ArrayList<>();
        Queue<Node> queue = new ArrayDeque<>();
        queue.add(root);
        int level = 0;
        while (!queue.isEmpty()) {
            int length = queue.size();
            while (length-- > 0) {
                Node curRoot = queue.poll();
                cache.computeIfAbsent(level, k -> new ArrayList<>()).add(curRoot.val);
                if (curRoot.children != null) {
                    for (Node child : curRoot.children) {
                        queue.add(child);
                    }
                }
            }
            level++;
        }
        for (int i = 0; i < cache.size(); i++) {
            res.add(cache.get(i));
        }
        return res;
    }
}




